src/core/prometheus.cc

```diff
@@ -198,33 +199,35 @@ static void fill_metric(pm::MetricFamily& mf, const metrics::impl::metric_value&
     }
 }
 
-static std::string to_str(seastar::metrics::impl::data_type dt) {
+static std::ostream& operator<<(std::ostream& os, seastar::metrics::impl::data_type dt) {
     switch (dt) {
     case seastar::metrics::impl::data_type::GAUGE:
-        return "gauge";
+        return os << "gauge";
     case seastar::metrics::impl::data_type::COUNTER:
     case seastar::metrics::impl::data_type::REAL_COUNTER:
-        return "counter";
+        return os << "counter";
     case seastar::metrics::impl::data_type::HISTOGRAM:
-        return "histogram";
+        return os << "histogram";
     case seastar::metrics::impl::data_type::SUMMARY:
-        return "summary";
+        return os << "summary";
     }
-    return "untyped";
+    return os << "untyped";
 }
 
-static std::string to_str(const seastar::metrics::impl::metric_value& v) {
+static std::ostream& operator<<(std::ostream& os, const seastar::metrics::impl::metric_value& v) {
     switch (v.type()) {
     case seastar::metrics::impl::data_type::GAUGE:
     case seastar::metrics::impl::data_type::REAL_COUNTER:
-        return std::to_string(v.d());
+        fmt::print(os, "{:.6f}", v.d());
```

Comment:
I don't think .6f is a good format. For small values, it loses precision. For large values, (beyond 10^(std::numeric_limit::digits10-6)), it has precision without accuracy.

I expect '{g}' will work and provide maximum accuracy, but haven't tested it.

Reply:
Should changing the value's precision be part of a patch that's supposedly about the labels?

I had in the past some bad experience (https://github.com/scylladb/scylladb/issues/19081#issuecomment-2147839659) with how Seastar's metrics convert double-precision to 6 digits of precision in obscure places, causing surprising effects. We need to be very careful about this. Note that using "to_string()" is even worse, because nobody knows what it really does... If I remember correctly (but I can't find a reference now), we even had a case where in the past some to-string function used to print 15 digits of precision but one day when we changed the way we use `fmt` suddenly it because 6 digits.

Ideally, we'd want to have regression tests for these precision issues, so we won't be surprised if one day the precision of some metric format regresses.

P.S. @avikivity , actually the default precision of printf's `%g` is 6 digits, similar to `%f` - the "g" is about scientific notation for large exponents - not about precision. But I don't know if fmt inherited printf's idea of what "g" is or invented its own - it's worth checking.

Reply:
This patch is already changing the precision, by changing to_string() to fmt::print("{:.6f").

If plain {g} doesn't work, then {:19g} (or whatever) can be substituted. Or even start with {} to maintain the status quo and improve it later.

Reply:
@avikivity this is exactly my question - why does a patch for "sanitize label value for text protocol" need to change the formatting of numbers.

Reply:
> This patch is already changing the precision

My intention with this PR is to keep the existing formatting of floating point numbers, keeping this PR purely a refactoring when it comes to the formatting of the metric values. The goal is to avoid the coming changes to `std::to_string`. `std::to_string` effectively used `{:.6f}` for the formatting of `double`s and this is what I intend to preserve with this change.

From https://en.cppreference.com/w/cpp/string/basic_string/to_string:

> 7,8) Converts a floating point value to a string as if by std::sprintf(buf, "%f", value). (until C++26)

Then from https://en.cppreference.com/w/cpp/io/c/fprintf:

> f/F: Converts floating-point number to the decimal notation in the style [-]ddd.ddd.
Precision specifies the exact number of digits to appear after the decimal point character. The default precision is 6. In the alternative implementation decimal point character is written even if no digits follow it. For infinity and not-a-number conversion style see notes.

I'm happy to move this commit to another PR if you prefer.

I think keeping the metric formatting as the original `std::to_string`-like format (`{:.6f}`) would be best. An important edge case for which `{:.6f}` works but `{g}` or `{}` would be problematic is gauges that have large integer values (eg. a gauge with a unix timestamp as its value). Changing the formatting to scientific notation could make these integer values less accurate which would be confusing.

Reply:
> > This patch is already changing the precision
> 
> My intention with this PR is to keep the existing formatting of floating point numbers, keeping this PR purely a refactoring when it comes to the formatting of the metric values. The goal is to avoid the coming changes to `std::to_string`. `std::to_string` effectively used `{:.6f}` for the formatting of `double`s and this is what I intend to preserve with this change.


Absolutely the correct approach.

> 
> From https://en.cppreference.com/w/cpp/string/basic_string/to_string:
> 
> > 7,8) Converts a floating point value to a string as if by std::sprintf(buf, "%f", value). (until C++26)
> 
> Then from https://en.cppreference.com/w/cpp/io/c/fprintf:
> 
> > f/F: Converts floating-point number to the decimal notation in the style [-]ddd.ddd.
> > Precision specifies the exact number of digits to appear after the decimal point character. The default precision is 6. In the alternative implementation decimal point character is written even if no digits follow it. For infinity and not-a-number conversion style see notes.
> 
> I'm happy to move this commit to another PR if you prefer.
> 
> I think keeping the metric formatting as the original `std::to_string`-like format (`{:.6f}`) would be best. An important edge case for which `{:.6f}` works but `{g}` or `{}` would be problematic is gauges that have large integer values (eg. a gauge with a unix timestamp as its value). Changing the formatting to scientific notation could make these integer values less accurate which would be confusing.

Let's plan a follow-up to improve floating-point formatting. .6f is pretty bad. It's okay for a gauge, but as soon as you take a derivative of a small slowly-changing metric, you get poor results.